### PR TITLE
[CF-1000] Sends up price and currency for inapps

### PR DESCRIPTION
**Pull request by joshdholtz**
### Checklist
- [x] If applicable, unit tests
- [x] If applicable, create follow-up issues for `purchases-ios` and hybrids

### Motivation

[CF-1000](https://revenuecats.atlassian.net/browse/CF-1000)

However, this PR alone won't make this work. This requires backend changes from [CF-1066](https://revenuecats.atlassian.net/browse/CF-1066).

### Description

- `ReceiptInfo` was default `price` and `currency` to `null`
  - In v5.6, these values were being defaulted to price and currency from a `StoreProduct`
  - This change brings that back ☝️ 
- `Purchases.postPurchases` was sending a `null` value for `StoreProduct` if no pricing phases were found
  - Will now find a matching `StoreProduct` for the first product identifier for use in `ReceiptInfo`
- Cleaned up `mockQueryingProductDetails()` in tests to make it easier to read for what it was doing different with `INAPP` and `SUBS`

#### Bonus
-  Deleted `.penaltyDeath()` from `MainApplication.kt` which was randomly crashing the app after like 20 seconds

#### Purchase tester request from Kibana

Screen Shot 2022-12-21 at 3 03 17 PM

### CODE REVIEW

File: examples/purchase-tester/src/main/java/com/revenuecat/purchasetester/MainApplication.kt [hunk 1]
```diff
@@ -29,7 +29,6 @@ class MainApplication : Application(), UpdatedCustomerInfoListener {
                 VmPolicy.Builder()
                     .detectLeakedClosableObjects()
                     .penaltyLog()
-                    .penaltyDeath()
```
Comment (joshdholtz) [inline examples/purchase-tester/src/main/java/com/revenuecat/purchasetester/MainApplication.kt:32]: Intentional deletion. This was causing random crashes that made testing purchase tester almost impossible.

File: common/src/main/java/com/revenuecat/purchases/common/ReceiptInfo.kt [hunk 2]
```diff
@@ -2,16 +2,16 @@ package com.revenuecat.purchases.common
...
 import com.revenuecat.purchases.models.StoreProduct
+import com.revenuecat.purchases.models.googleProduct
 
 class ReceiptInfo(
     val productIDs: List<String>,
     val offeringIdentifier: String? = null,
     val purchaseOptionId: String? = null,
     val storeProduct: StoreProduct? = null,
 
-    // TODO BC5 - replace price/currency with making a single PricingPhase for amazon observermode products?
```
Comment (joshdholtz) [inline common/src/main/java/com/revenuecat/purchases/common/ReceiptInfo.kt:12]: I think this is okay to delete this comment? 🤔 

The code changed below is almost example what v5.6 used which should still work for Amazon? (Hopefully 😁)

File: purchases/src/test/java/com/revenuecat/purchases/PurchasesTest.kt [hunk 3]
```diff
@@ -4353,17 +4353,22 @@ class PurchasesTest {
...
+            )
 
-        return mockQueryingProductDetails(storeProduct, offeringIdentifier)
+            return mockQueryingProductDetails(storeProduct, offeringIdentifier)
+        } else {
+            val productDetails = createMockOneTimeProductDetails(productId, 2.00)
+            val storeProduct = productDetails.toInAppStoreProduct()
+
+            return mockQueryingProductDetails(storeProduct, offeringIdentifier)
+        }
```
Comment (joshdholtz) [inline purchases/src/test/java/com/revenuecat/purchases/PurchasesTest.kt:4371]: The inline if statements made this a little hard to follow and debug so split it up into two clear paths for `SUBS` and `INAPP`

File: purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt [hunk 4]
```diff
@@ -1235,15 +1253,10 @@ class Purchases internal constructor(
...
                     )
-                } else {
-                    postToBackend(
-                        purchase = purchase,
-                        storeProduct = null,
-                        allowSharingPlayStoreAccount = allowSharingPlayStoreAccount,
-                        consumeAllTransactions = consumeAllTransactions,
-                        appUserID = appUserID,
-                        onSuccess = onSuccess,
-                        onError = onError
```
[the remark this answers is not in the document]
Comment (joshdholtz) [reply purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt:1246]: `price` and `currency` _do_ get left out when `storeProduct` is `null`

And yeah, that will happen on a restore still. Here is what a restore looked like 👇 

Screen Shot 2022-12-22 at 7 08 35 AM

File: examples/purchase-tester/src/main/java/com/revenuecat/purchasetester/MainApplication.kt (hunk 1 above)
Comment (joshdholtz) [reply examples/purchase-tester/src/main/java/com/revenuecat/purchasetester/MainApplication.kt:32]: Thanks for pointing that out the other day 💪

File: purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (swehner) [reply purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt:1246]: So, then we should be keeping the same behavior, right?

[the remark this answers is not in the document]
Comment (joshdholtz) [reply purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt:1246]: Yes yes! Restores will keep the same behavior as before (no `price` or `currency`). New purchases will send up `price` and `currency`. That is what we want, right? 🤷‍♂️

File: common/src/main/java/com/revenuecat/purchases/common/ReceiptInfo.kt (hunk 2 above)
Comment (beylmk) [reply common/src/main/java/com/revenuecat/purchases/common/ReceiptInfo.kt:12]: i think we can remove it. we'll have to retest alllll the things for amazon anyway...

Comment (beylmk) [reply common/src/main/java/com/revenuecat/purchases/common/ReceiptInfo.kt:12]: although it is interesting that i noted those were only passed for amazon observer mode.. how did backend figure out price/currency before?

File: examples/purchase-tester/src/main/java/com/revenuecat/purchasetester/MainApplication.kt (hunk 1 above)
Comment (beylmk) [reply examples/purchase-tester/src/main/java/com/revenuecat/purchasetester/MainApplication.kt:32]: @vegaro any context on why that was there?

File: common/src/main/java/com/revenuecat/purchases/common/ReceiptInfo.kt [hunk 5]
```diff
@@ -9,9 +9,8 @@ class ReceiptInfo(
...
-    val price: Double? = null, // only passed for amazon observer mode
-    val currency: String? = null // only passed for amazon observer mode
+    val price: Double? = storeProduct?.oneTimeProductPrice?.priceAmountMicros?.div(MICROS_MULTIPLIER.toDouble()),
+    val currency: String? = storeProduct?.oneTimeProductPrice?.currencyCode
```
Comment (beylmk) [inline common/src/main/java/com/revenuecat/purchases/common/ReceiptInfo.kt:13]: might be good to add a test to confirm the logic in these...i think right now we just test that the receipt info passed in == the receipt info posted, which contains these values. but i'm thinking more something that confirms the currency is in fact the onetimeproductprice's currency

File: purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (swehner) [reply purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt]: Thanks for putting it back :)

File: purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt [hunk 6]
```diff
@@ -1235,7 +1241,7 @@ class Purchases internal constructor(
                             )
                         }
                     )
-                } else {
+                } ?: run {
```
Comment (tonidero) [inline purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt:1244]: Hmm I'm wondering, will this case ever happen? As in, having a `StoreTransaction` without any `productId`... Seems like a bug to me and I would probably log an error, but not sure if I'm missing some valid case here.

File: purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt [hunk 7]
```diff
@@ -1202,15 +1202,21 @@ class Purchases internal constructor(
     ) {
         purchases.forEach { purchase ->
             if (purchase.purchaseState != PurchaseState.PENDING) {
-                if (purchase.purchaseOptionId != null) {
+                purchase.productIds.firstOrNull()?.let { productId ->
                     billing.queryProductDetailsAsync(
                         productType = purchase.type,
-                        productIds = purchase.productIds.toSet(),
+                        productIds = setOf(productId),
```
Comment (tonidero) [inline purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt:1208]: Hmm is there any reason for this change? I think purchases should only have a single productId for now (until multi-line purchases maybe?) but we can still query all productIDs and just use the first store product inside the completion block, same as before. I might be missing something though.

File: common/src/main/java/com/revenuecat/purchases/common/ReceiptInfo.kt (hunk 5 above)
Comment (tonidero) [reply common/src/main/java/com/revenuecat/purchases/common/ReceiptInfo.kt:13]: Oh I'm rethinking this, when syncing observer mode amazon purchases, these could be the price/currency of a subscription I guess... But then, we probably should try to map that info into the `PricingPhases` class? Though that would not be possible without more info 🤔. I need to think more about this...

File: purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt (hunk 6 above)
Comment (joshdholtz) [reply purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt:1244]: @tonidero Ha! I had an error here originally in this PR but then it looked like it changed the behavior as @swehner mentioned so I removed it 🤷‍♂️ 

I think the concern was with restores which is why I put this back to the way it is now 🤔

File: purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt (hunk 7 above)
Comment (joshdholtz) [reply purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt:1208]: Ummm... I think you right! I'll revert this since this could get missed in the future when we do need to support multiple

File: purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt (hunk 6 above)
Comment (tonidero) [reply purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt:1244]: Oh, I think this code does not get called for restores, but it does get called for pending purchases. But in both cases, I think we should still have a `productId` so I think this code shouldn't be reached... Again, not sure if I'm missing some case where this could be null but in `main`, it doesn't seem like that's a case we are handling right now https://github.com/RevenueCat/purchases-android/blob/main/purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt#L1153-L1178

Comment (joshdholtz) [reply purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt:1244]: So should we changes this to verify `purchase.productIds` is not empty before trying to query store products? And if empty then throw an error?

Comment (tonidero) [reply purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt:1244]: Well, it looks like we were not verifying that here at all in 5.6.x. Instead, when querying the productIds with Google, we would show a debug log: https://github.com/RevenueCat/purchases-android/blob/main/feature/google/src/unityIAP/java/com/revenuecat/purchases/google/BillingWrapper.kt#L149. 

So as long as we revert [this other change](https://github.com/RevenueCat/purchases-android/pull/721#discussion_r1055659367), IMO, we should just go back to that and not verify that case here at all.

File: purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt (hunk 7 above)
Comment (beylmk) [reply purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt:1208]: i think it might be safer to get the single product Id here, otherwise we query for multiple store products. hopefully we only pass the single ID higher up in the call chain, but this is where it really matters

Comment (beylmk) [reply purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt:1208]: though that is a functionality change so i'm okay if we want to avoid those for now...

Comment (tonidero) [reply purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt:1208]: Well, right now it shouldn't really matter, since purchases should only have a single product ID. But I guess I would prefer to minimize the changes unless there is a good reason. 

Also, if devs start using multi-line purchases without proper support, I wouldn't focus in optimizing, but rather on making sure they know that's not supported. As in, if we detect a purchase with multiple product ids, I would prefer to fail rather than just pick the first and continue the process, and we should fail as early as possible.

File: common/src/main/java/com/revenuecat/purchases/common/ReceiptInfo.kt (hunk 5 above)
Comment (joshdholtz) [reply common/src/main/java/com/revenuecat/purchases/common/ReceiptInfo.kt:13]: @tonidero Why would Amazon observer mode subscriptions fall under the `oneTimeProductPrice` property? 🤔 I don't think I follow that 😅

File: purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt (hunk 7 above)
Comment (beylmk) [reply purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt:1208]: okay sure that makes sense to me. if we go that route maybe we can make a new ticket to handle those failures

File: common/src/main/java/com/revenuecat/purchases/common/ReceiptInfo.kt (hunk 2 above)
Comment (beylmk) [reply common/src/main/java/com/revenuecat/purchases/common/ReceiptInfo.kt:12]: ohhh it'd be in the StoreProduct....my brain is mush sorry. either way, i think it's okay to delay amazon stuff for later.

as long as backend knows these values will be coming in for OTP now, and not just amazon observer mode

File: common/src/main/java/com/revenuecat/purchases/common/ReceiptInfo.kt (hunk 5 above)
Comment (beylmk) [reply common/src/main/java/com/revenuecat/purchases/common/ReceiptInfo.kt:13]: i _think_ i might be following. when we sync amazon observer mode purchases, we don't have a way of getting price/currency ourselves, so we require the user to pass it in. that's what these fields used to be for.

for any non-amazon-observermode purchases, we would just pass the price/currency in the storeproduct.

now, our backend is repurposing the price/currency field to be used for OTP (right?) so now we have a need for those to work for amazon observer mode purchases (both subs and inapps), AND for all other inapp purchases.

does that make sense?

Comment (beylmk) [reply common/src/main/java/com/revenuecat/purchases/common/ReceiptInfo.kt:13]: so we could rename these to be the onetimeProductPrice/Currency, but we still need a field for amazon observer mode products (potentially subs)

Comment (beylmk) [reply common/src/main/java/com/revenuecat/purchases/common/ReceiptInfo.kt:13]: so we need to think through how amazon observer mode subs pricing info would get passed.

okay those comments were just rambles, that took me some time to figure out. maybe that helps you understand hearing it differently @joshdholtz ?

Comment (tonidero) [reply common/src/main/java/com/revenuecat/purchases/common/ReceiptInfo.kt:13]: So, as part of the SDK API, we have a method [`syncObserverModeAmazonPurchase`](https://github.com/RevenueCat/purchases-android/blob/main/purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt#L266). That's used so devs can give us price data for their purchases, subscriptions or not. Currently, we just store the price and currency parameters from that method in the `ReceiptInfo` `price` and `currency` fields. 

If we separate the price info between `oneTimeProductPrice` and `subscriptionPricingPhases` depending on the type of purchase, we wouldn't know how to store the data given by devs in `syncObserverModeAmazonPurchase` unless we ask devs for it.

Not sure if there is a different way to find that info from the SDK in observer mode though... I'm not that familiar with the Amazon store.

Comment (tonidero) [reply common/src/main/java/com/revenuecat/purchases/common/ReceiptInfo.kt:13]: And yeah @beylmk's rambles are correct 😛 . Sorry I missed some steps on my reasoning there 😅

File: purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt (hunk 6 above)
Comment (joshdholtz) [reply purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt]: Reverted this back to sending up the entire list/set of product ids!

File: common/src/main/java/com/revenuecat/purchases/common/ReceiptInfo.kt (hunk 5 above)
Comment (beylmk) [reply common/src/main/java/com/revenuecat/purchases/common/ReceiptInfo.kt:13]: so it feels like we should at least think about whether amazon pricing phases will work before finalizing this OTP functionality. but i think we always knew amazon might cause some changes, so i'm OK merging this PR for now...

Comment (joshdholtz) [reply common/src/main/java/com/revenuecat/purchases/common/ReceiptInfo.kt:13]: Okay, so... I added some new tests to `ReceiptInfo` for when `price` and `currency` are manually passed in and it seems like what I have in this PR should work for Amazon still 🤞 

[syncObserverModeAmazonPurchase](https://github.com/RevenueCat/purchases-android/blob/main/purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt#L266) doesn't pass in a `storeProduct` but only `productIds`, `price`, and `currency`. 

So I _tthhiiinnnnkkk_ this should be okay for all the use cases now? 😇

File: common/src/main/java/com/revenuecat/purchases/common/ReceiptInfo.kt (hunk 2 above)
Comment (joshdholtz) [reply common/src/main/java/com/revenuecat/purchases/common/ReceiptInfo.kt:12]: Ahhh yeah, @greenietea is making changes for the backend to expect price and currency for Google Play OTP!

v5.6 did pass up price and currency so this was a breaking change early in in v6 I think? 🤔

File: common/src/main/java/com/revenuecat/purchases/common/ReceiptInfo.kt (hunk 5 above)
Comment (beylmk) [reply common/src/main/java/com/revenuecat/purchases/common/ReceiptInfo.kt:13]: i'm good with it! i'll link this PR in our amazon tickets so we remember this conversation

Comment (tonidero) [reply common/src/main/java/com/revenuecat/purchases/common/ReceiptInfo.kt:13]: Yeah, I think functionally this works. My only concern is about these fields being confusing with `pricingPhases`. I don't think that's a new problem from this PR so don't want to block this just because of that but I would add a comment or create a ticket to address in the future.

File: examples/purchase-tester/src/main/java/com/revenuecat/purchasetester/MainApplication.kt (hunk 1 above)
Comment (tonidero) [reply examples/purchase-tester/src/main/java/com/revenuecat/purchasetester/MainApplication.kt:32]: I've had this happen a few times... Usually uninstalling/reinstalling fixes the crashes for me. It would be great to understand the issue but yeah, I'm not sure it's worth to keep it

**Review by tonidero — APPROVED** [review]
LGTM!

File: common/src/main/java/com/revenuecat/purchases/common/ReceiptInfo.kt (hunk 5 above)
Comment (swehner) [reply common/src/main/java/com/revenuecat/purchases/common/ReceiptInfo.kt:13]: BTW: Yes, it looks like for amazon receipts we are persisting the price sent in and use it.

**Review by swehner — APPROVED** [review]

File: common/src/main/java/com/revenuecat/purchases/common/ReceiptInfo.kt (hunk 5 above)
Comment (beylmk) [reply common/src/main/java/com/revenuecat/purchases/common/ReceiptInfo.kt:13]: so we'll just need to make sure that we either add a different field or that backend has some way of differentiating between amazon prices vs google OTPS^^

@greenietea FYI

Comment (joshdholtz) [reply common/src/main/java/com/revenuecat/purchases/common/ReceiptInfo.kt:13]: > some way of differentiating between amazon prices vs google OTPS

We should be able to use the platform header for this, right?

Comment (beylmk) [reply common/src/main/java/com/revenuecat/purchases/common/ReceiptInfo.kt:13]: yeah, i'd think so!
